### PR TITLE
Guard raw refreshes against incoherent coordinate shifts

### PR DESCRIPTION
**Pull request by michaelmwu**
## Summary

- preserve prior coordinates when a refreshed record keeps a strong identity and stable address but moves more than 1 km or loses half of its coordinate pair
- keep normal coordinate jitter and relocations with a changed address
- compare international addresses with Unicode-safe normalization and keep Maps URLs only when their embedded identity is compatible

## Why

Google saved-list payloads can temporarily mix a stable place identity and old address with stale or future coordinates. That surfaced as 5.23 km and 2.71 km marker jumps in [demflyers-favorites PR #101](https://github.com/michaelmwu/demflyers-favorites/pull/101).

## Verification

- `bun run test` — 137 frontend and 658 Python tests passed
- `bun run check`
- `bun run build`
- `uv run ruff check scripts/build_data.py tests/python/test_build_data.py`
- `git diff --check`

---

> [!NOTE]
> **Medium Risk**
> Changes core raw-place merge, coordinate, and Google identity heuristics in the data build path; mistakes could mis-merge places or leave bad map positions, but behavior is heavily regression-tested.
> 
> **Overview**
> **Hardens raw saved-list refresh merging** so stale Google payloads do not overwrite good coordinates, addresses, or Maps links when the place is still the same.
> 
> When refreshed data keeps a **compatible identity** and a **stable street-level address** but lat/lng jump more than **1 km** (or one coordinate is missing), `preserve_existing_raw_place` now keeps the prior coordinates and can also keep address, `google_id`/`cid`/`maps_place_token`, and `maps_url` when URL embedded IDs still match. Address “stability” and precision regression use new **international address normalization** (aliases, ordinals, postal/subdivision/country handling, unit stripping, etc.). Name matching is tightened (e.g. rejects generic `item` slugs).
> 
> **Identity and dedup:** `query_place_id` from Maps URLs is treated as **`gpid`** in stable IDs and match keys. Duplicate **CID** and **Google identity** cleanup now considers identities from URLs and fields, runs in a **loop** until stable, prefers consistent single-identity rows, and rebuilds Maps URLs via **`build_raw_place_maps_url_preserving_identity`** instead of always falling back to generic search links. A **fallback matcher** can re-link a refreshed name-only row to a single prior row that still has durable Google identity and stable address.
> 
> **Pipeline order:** duplicate clearing runs on the refreshed payload **before** the suspicious identity-loss guard and per-place preservation merge.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 03a4a6724ee72bfabdaa8d391461601dd254febd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved place refresh merging to better preserve existing coordinates and address details when the refreshed data clearly refers to the same location, including safeguards against precision/format regressions.
  * More reliably preserved Google Maps links by tightening identity matching and resolving duplicates without causing unintended identity changes.
  * Enhanced duplicate cleanup so place identity signals are deduplicated consistently during refresh/merge.
* **Tests**
  * Expanded regression tests to cover coordinate/address preservation and Google Maps link handling across many refresh and deduplication edge cases.

### CODE REVIEW

File: scripts/build_data.py [hunk 1]
```diff
@@ -6715,6 +6716,19 @@ def preserve_existing_raw_place(
     if names_compatible and refreshed_place.added_by is None and existing_place.added_by is not None:
         updates["added_by"] = existing_place.added_by
         preserved_fields.append("added_by")
+    if names_compatible and raw_place_coordinates_should_be_preserved(
+        existing_place,
+        refreshed_place,
+    ):
+        updates["lat"] = existing_place.lat
+        updates["lng"] = existing_place.lng
+        preserved_fields.append("coordinates")
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:6899]: Fixed in 0eb0297. Coordinate preservation now relies on the stable Google identity/CID and address checks even when the display name changes, with a regression covering a complete rename.

File: scripts/build_data.py [hunk 2]
```diff
@@ -6715,6 +6716,19 @@ def preserve_existing_raw_place(
     if names_compatible and refreshed_place.added_by is None and existing_place.added_by is not None:
         updates["added_by"] = existing_place.added_by
         preserved_fields.append("added_by")
+    if names_compatible and raw_place_coordinates_should_be_preserved(
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 0eb0297: the outer name gate is removed, and the regression verifies a complete display-name change with stable CID/address still preserves the prior coordinates.

File: scripts/build_data.py [hunk 3]
```diff
@@ -6732,6 +6762,167 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+            token
+            for token in shorter_tokens
+            if any(character.isdigit() for character in token)
+        }
+        if (
+            len(shorter_tokens) >= 3
+            and numeric_tokens
+            and shorter_tokens <= longer_tokens
+        ):
+            return True
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7354]: Fixed in 09024e4. Street anchors now come from ordered road-segment tokens, excluding immediately associated unit, suite, floor, building, and tower numbers; Australian slash-unit notation is handled explicitly. Regressions cover inline unit/street swaps and 5/12 Main St versus Unit 5, 12 Main Street.

File: scripts/build_data.py [hunk 4]
```diff
@@ -6732,6 +6762,167 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+
+def raw_place_maps_url_should_be_preserved(
+    existing_place: RawPlace,
+    refreshed_place: RawPlace,
+) -> bool:
+    existing_url = as_string(existing_place.maps_url)
+    refreshed_url = as_string(refreshed_place.maps_url)
+    if not existing_url or existing_url == refreshed_url:
+        return False
+    if google_maps_uri_strength(existing_url) < google_maps_uri_strength(refreshed_url):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:8247]: Fixed in 8d0d3c5. A refreshed Maps URL that exposes CID, query_place_id, or a 0x place token now outranks an existing URL with no embedded identity even when their coarse URI strengths tie; the equal-strength opaque-place versus CID regression covers this case.

File: scripts/build_data.py [hunk 5]
```diff
@@ -6732,6 +6762,167 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        return False
+    if existing_address_key == refreshed_address_key:
+        return True
+
+    existing_street_numbers = address_street_numbers(existing_address)
+    refreshed_street_numbers = address_street_numbers(refreshed_address)
+    if (
+        existing_street_numbers
+        and refreshed_street_numbers
+        and not existing_street_numbers & refreshed_street_numbers
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 09024e4. Coordinate address matching now extracts ordered street anchors, excludes associated unit numbers, requires equal street-number sets, and handles slash-unit notation. Regressions cover 12 Main St versus 312 Main St Unit 12 and inline unit/street swaps.

File: scripts/build_data.py [hunk 6]
```diff
@@ -6732,6 +6784,270 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    existing_place: RawPlace,
+    refreshed_place: RawPlace,
+) -> bool:
+    existing_address = as_string(existing_place.address)
+    if not existing_address:
+        return False
+
+    refreshed_address = as_string(refreshed_place.address)
+    if not refreshed_address:
+        return raw_place_address_is_preservable(existing_place, refreshed_place)
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7063]: Fixed in fdc75e5. The missing-address path now uses the same strict street-level evidence gate before preserving coordinates. Locality-only prior values such as Paris, France no longer roll back a corrected pin; regression coverage also distinguishes postcode-only localities from tracked international street forms.

File: scripts/build_data.py [hunk 7]
```diff
@@ -6732,6 +6784,432 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        return False
+
+    existing_tokens, existing_country = raw_place_coordinate_address_comparison_key(
+        existing_address
+    )
+    refreshed_tokens, refreshed_country = raw_place_coordinate_address_comparison_key(
+        refreshed_address
+    )
+    if not existing_tokens or existing_tokens != refreshed_tokens:
+        return False
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7133]: Fixed in 0c4d2e1e. Address tokens now remain ordered within each comma component while the components themselves are compared canonically, so equivalent component reordering is stable without recreating the prior York/London token-collision bug. Postal-only tokens are normalized separately.

File: scripts/build_data.py [hunk 8]
```diff
@@ -6732,6 +6784,388 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    if existing_url_cid and refreshed_cids and existing_url_cid not in refreshed_cids:
+        return False
+
+    existing_query_place_id = extract_maps_query_place_id(existing_url)
+    refreshed_query_place_id = extract_maps_query_place_id(refreshed_url)
+    if (
+        existing_query_place_id
+        and refreshed_query_place_id
+        and existing_query_place_id != refreshed_query_place_id
+    ):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:8273]: Fixed in 0c4d2e1e. Maps URL preservation now requires at least one verifiable same-type identity when both URLs embed identities; a prior query_place_id can no longer overwrite a refreshed CID URL when those identities cannot be cross-checked.

File: scripts/build_data.py [hunk 9]
```diff
@@ -6732,6 +6784,432 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+
+
+def raw_place_coordinate_address_is_stable(
+    existing_place: RawPlace,
+    refreshed_place: RawPlace,
+) -> bool:
+    existing_address = as_string(existing_place.address)
+    if not existing_address:
+        return False
+    if not raw_place_coordinate_address_has_street_level_evidence(existing_address):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7058]: Fixed in 37c5ad12. Romanized Chome is recognized as street-level evidence; 1 Chome-63-1 Jingumae now has a regression test.

File: scripts/build_data.py [hunk 10]
```diff
@@ -6732,6 +6784,432 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
     return left_maps_place_token is not None and left_maps_place_token == right_maps_place_token
 
 
+def raw_place_coordinates_should_be_preserved(
+    existing_place: RawPlace,
+    refreshed_place: RawPlace,
+) -> bool:
+    if not raw_place_coordinate_identity_matches(existing_place, refreshed_place):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:6927]: Fixed in 37c5ad12. A compatible-name refresh that loses every durable ID can enter the coordinate guard when the existing row had durable identity; stable street address is still required, and any refreshed CID alias blocks the identity-loss fallback.

File: scripts/build_data.py [hunk 11]
```diff
@@ -6732,6 +6784,432 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+
+        street_suffix_indexes = {
+            index
+            for index, token in enumerate(tokens)
+            if token in RAW_PLACE_STREET_SUFFIX_TOKENS
+        }
+        if any(index > 0 for index in street_suffix_indexes):
+            return True
+        if set(tokens) & {"hwy", "rte"}:
+            return True
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 37c5ad12. Route-only forms remain movable, including Highway 1, Route 66, US Highway 1, State Route 66, M1 Highway, and unit-number variants. A real premise such as 123 US Highway 1 still qualifies.

File: scripts/build_data.py [hunk 12]
```diff
@@ -6732,6 +6784,432 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+            existing_lng,
+            refreshed_lat,
+            refreshed_lng,
+        )
+        > RAW_PLACE_SUSPICIOUS_COORDINATE_SHIFT_METERS
+    )
+
+
+def raw_place_coordinate_identity_matches(left: RawPlace, right: RawPlace) -> bool:
+    if raw_place_strong_google_identity_matches(left, right):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 37c5ad12. query_place_id now participates both in saved-list preservation match keys and coordinate identity matching, with a saved-list-level regression covering coordinate and cosmetic address changes.

File: scripts/build_data.py [hunk 13]
```diff
@@ -6732,6 +6804,546 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+def raw_place_coordinate_address_token_list(address_key: str) -> list[str]:
+    return [
+        RAW_PLACE_ADDRESS_TOKEN_ALIASES.get(token, token)
+        for token in address_key.split()
+    ]
+
+
+def raw_place_coordinate_address_has_street_level_evidence(address: str) -> bool:
+    normalized_address = raw_place_coordinate_strip_japanese_postal_prefix(address)
+    for part in re.split(r"[,、]", normalized_address):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7323]: Fixed in 37c5ad12. Adjacent premise-number components are now joined for street-evidence purposes; Piazza del Colosseo, 1 follows the same covered path as Via Roma, 1.

File: scripts/build_data.py [hunk 14]
```diff
@@ -6732,6 +6804,546 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+            if token in RAW_PLACE_STREET_SUFFIX_TOKENS
+        }
+        if any(index > 0 for index in street_suffix_indexes):
+            return True
+        if set(tokens) & {"hwy", "rte"}:
+            return True
+        if re.search(
+            r"\b(?:allee|allée|av|avenida|calle|carrer|chaussee|chaussée|chemin|corso|impasse|piazza|quai|rua|rue|strasse|straße|via)\b",
+            address_key,
+        ):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 84c219e7. Compound German street tokens ending in -straße or -strasse now count as street evidence; both Oranienstraße and Oranienstrasse are covered.

File: scripts/build_data.py [hunk 15]
```diff
@@ -6732,6 +6804,546 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+                comparison_part_tokens.append(token)
+                index += 1
+                continue
+
+            index = unit_identifier_index + 1
+        for country_suffix, country_code in raw_place_coordinate_country_token_identities():
+            suffix_length = len(country_suffix)
+            if tuple(comparison_part_tokens[-suffix_length:]) == country_suffix:
+                del comparison_part_tokens[-suffix_length:]
+                country = country_code
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 84c219e7. Countries are detected only in the terminal component. A nonterminal country token is stripped only when the peer address has already established that exact country; Mexico, MO remains locality evidence.

File: scripts/build_data.py [hunk 16]
```diff
@@ -6732,6 +6808,699 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+def raw_place_coordinate_identity_matches(left: RawPlace, right: RawPlace) -> bool:
+    if raw_place_strong_google_identity_matches(left, right):
+        return True
+
+    left_query_place_id = extract_maps_query_place_id(left.maps_url)
+    right_query_place_id = extract_maps_query_place_id(right.maps_url)
+    if (
+        left_query_place_id is not None
+        and left_query_place_id == right_query_place_id
+    ):
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 738c6720. Opaque stored Google Place IDs and query_place_id values share one consistency-checked identity set and preservation key; legacy g/ and /g/ IDs remain in their separate namespace.

File: scripts/build_data.py [hunk 17]
```diff
@@ -6732,6 +6808,699 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    if not candidate_indexes:
+        return set()
+
+    street_suffix_index = max(
+        index
+        for index, token in enumerate(tokens)
+        if token in RAW_PLACE_STREET_SUFFIX_TOKENS
+    )
+    before_suffix = [index for index in candidate_indexes if index < street_suffix_index]
+    selected_indexes = before_suffix or [
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:8219]: Fixed in 738c6720. Numeric and written ordinals normalize to the same token without counting as premise numbers. Two ordinal street pairs still qualify as a precise intersection.

File: scripts/build_data.py [hunk 18]
```diff
@@ -6732,6 +6808,699 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        return False
+    return bool(
+        existing_country is None
+        or refreshed_country is None
+        or existing_country == refreshed_country
+    )
+
+
+def raw_place_coordinate_address_key(address: str) -> str | None:
+    normalized = unicodedata.normalize("NFKC", address).casefold()
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 738c6720. Address keys fold Latin diacritics while retaining non-Latin combining marks; Colón/Mérida and a Devanagari non-collision are covered.

File: scripts/build_data.py [hunk 19]
```diff
@@ -159,6 +159,44 @@ def update(self, **_: Any) -> None:
 PHOTOLESS_REAL_PLACE_CACHE_TTL = timedelta(days=1)
 RAW_SOURCE_CACHE_TTL = timedelta(days=14)
 RAW_SOURCE_REFRESH_JITTER = timedelta(days=3)
+RAW_PLACE_SUSPICIOUS_COORDINATE_SHIFT_METERS = 1_000.0
+RAW_PLACE_ADDRESS_TOKEN_ALIASES = {
+    "avenue": "ave",
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:166]: Fixed in 738c6720. Localized aliases now normalize Av./Avenida and the covered R./Rua and Str./Strasse forms before structural comparison.

File: scripts/build_data.py [hunk 20]
```diff
@@ -6732,6 +6808,699 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+            if (
+                unit_identifier_index < len(part_tokens)
+                and part_tokens[unit_identifier_index]
+                in RAW_PLACE_UNIT_NUMBER_LABEL_TOKENS
+            ):
+                unit_identifier_index += 1
+            is_unit_designator = bool(
+                token in RAW_PLACE_UNIT_PREFIX_TOKENS
+                and unit_identifier_index < len(part_tokens)
+                and part_tokens[unit_identifier_index]
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7772]: Fixed in 738c6720. Trailing floor/level designators such as 2nd Floor are removed symmetrically with prefix forms such as Floor 2.

File: scripts/build_data.py [hunk 21]
```diff
@@ -6732,6 +6808,755 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+                comparison_part_tokens,
+                index,
+                street_numbers,
+            )
+        ]
+        if comparison_part_tokens:
+            comparison_parts.append(tuple(comparison_part_tokens))
+    subdivision_identities = raw_place_coordinate_subdivision_token_identities(
+        country or country_hint
+    )
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7847]: Fixed in 4d267c20. A globally canonical full subdivision name can infer one country hint when both addresses omit country, allowing California/CA while refusing ambiguous multi-country evidence.

File: scripts/build_data.py [hunk 22]
```diff
@@ -6732,6 +6808,755 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        return True
+
+    return bool(
+        raw_place_names_are_compatible(
+            as_string(left.name),
+            as_string(right.name),
+        )
+        and raw_place_has_durable_google_identity(left)
+        and not raw_place_has_durable_google_identity(right)
+    )
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7038]: Fixed in 4d267c20. Name compatibility first compares Unicode-normalized keys and rejects the generic ASCII slug item, so unrelated non-ASCII names cannot enter identity-loss preservation.

File: scripts/build_data.py [hunk 23]
```diff
@@ -6732,6 +6808,755 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    ...,
+]:
+    country_alias_codes = {
+        "England": "GB",
+        "Scotland": "GB",
+        "Wales": "GB",
+        "Northern Ireland": "GB",
+        "UK": "GB",
+        "UAE": "AE",
+        "USA": "US",
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:8011]: Fixed in 4d267c20. Terminal US now canonicalizes to the same US country identity as USA and United States.

File: scripts/build_data.py [hunk 24]
```diff
@@ -6732,6 +6863,903 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        )
+        > RAW_PLACE_SUSPICIOUS_COORDINATE_SHIFT_METERS
+    )
+
+
+def raw_place_coordinate_identity_matches(left: RawPlace, right: RawPlace) -> bool:
+    left_google_places_ids = raw_place_google_places_identities(left)
+    right_google_places_ids = raw_place_google_places_identities(right)
+    if len(left_google_places_ids) > 1 or len(right_google_places_ids) > 1:
+        return False
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in f6ee69fa. Matching primary CIDs can now protect coordinates even when a stored/query Place ID pair is internally inconsistent; alias-only CID overlap requires compatible names plus coherent, nonconflicting structured IDs. Added positive primary/alias and negative conflict regressions.

File: scripts/build_data.py [hunk 25]
```diff
@@ -6537,6 +6619,10 @@ def raw_place_primary_match_keys(place: RawPlace, *, source_type: str | None = N
     if maps_place_token:
         keys.append(f"gms:{maps_place_token}")
 
+    google_places_id = raw_place_google_places_id_identity(place)
+    if google_places_id:
+        keys.append(f"gpid:{google_places_id}")
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:6638]: Fixed in f6ee69fa. CID, Google Place ID, and map-token cleanup now indexes all field and URL occurrences, prefers internally coherent keepers, and iterates to a fixed point before preservation matching. Duplicate query IDs are removed without dropping independent CIDs, tokens, or Place IDs; cross-form, reverse-order, and compound-URL regressions cover this.

File: scripts/build_data.py [hunk 26]
```diff
@@ -6732,6 +6863,903 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        if comparison_part_tokens:
+            comparison_parts.append(tuple(comparison_part_tokens))
+    subdivision_identities = raw_place_coordinate_subdivision_token_identities(
+        country or country_hint
+    )
+    comparison_parts = [
+        subdivision_identities.get(part, part)
+        for part in comparison_parts
+    ]
+    return tuple(sorted(comparison_parts)), country
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7848]: Fixed in f6ee69fa. Embedded subdivision codes are split with country-valid postal context, including Boston MA, Toronto ON, Orlando FL, and postal-led Italian forms. The normalizer also avoids unit/name false positives such as Floor, Room, Suite, Hotel California, and Made In. Added international and punctuation-variation regressions.

File: scripts/build_data.py [hunk 27]
```diff
@@ -6732,12 +6867,944 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        previous_base_is_latin = bool(
+            category.startswith("L")
+            and "LATIN" in unicodedata.name(character, "")
+        )
+    normalized = unicodedata.normalize("NFKC", "".join(without_latin_diacritics))
+    normalized_characters = [
+        character
+        if unicodedata.category(character)[0] in {"L", "M", "N"}
+        else " "
+        for character in normalized
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7202]: Fixed in 94768921. In-word straight and typographic apostrophes are normalized consistently, with a King’s/Kings address regression test.

File: scripts/build_data.py [hunk 28]
```diff
@@ -6732,12 +6867,944 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+]:
+    country_alias_codes = {
+        "England": "GB",
+        "Scotland": "GB",
+        "Wales": "GB",
+        "Northern Ireland": "GB",
+        "UK": "GB",
+        "UAE": "AE",
+        "US": "US",
+        "USA": "US",
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:8011]: Fixed in 94768921. Dotted U.S. and U.S.A. suffixes now normalize to the same US country identity, with regression coverage for both forms.

File: scripts/build_data.py [hunk 29]
```diff
@@ -6732,12 +6867,944 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    return token.startswith("ordinal_")
+
+
+def raw_place_coordinate_part_has_street_intersection(tokens: list[str]) -> bool:
+    return sum(
+        raw_place_coordinate_token_is_ordinal(token)
+        and index + 1 < len(tokens)
+        and tokens[index + 1] in RAW_PLACE_STREET_SUFFIX_TOKENS
+        for index, token in enumerate(tokens)
+    ) >= 2
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7317]: Fixed in 94768921 and tightened in e6d85d19. Named intersections with an explicit &, and, or at connector are now street-level evidence; the two-ordinal behavior remains. A negative One Court Square regression prevents suffix-like building names from being misclassified.

File: scripts/build_data.py [hunk 30]
```diff
@@ -159,6 +159,88 @@ def update(self, **_: Any) -> None:
 PHOTOLESS_REAL_PLACE_CACHE_TTL = timedelta(days=1)
 RAW_SOURCE_CACHE_TTL = timedelta(days=14)
 RAW_SOURCE_REFRESH_JITTER = timedelta(days=3)
+RAW_PLACE_SUSPICIOUS_COORDINATE_SHIFT_METERS = 1_000.0
+RAW_PLACE_ADDRESS_TOKEN_ALIASES = {
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:163]: Fixed in 386ac6bb. Full compass words now normalize to N/S/E/W and diagonal abbreviations in a separate directional alias map, so they do not accidentally become street-suffix markers. Added the Pennsylvania Ave NW/Northwest regression.

File: scripts/build_data.py [hunk 31]
```diff
@@ -6732,175 +6867,1533 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        character.isdigit() for character in premise_tokens[0]
+    ):
+        return False
+    for token in premise_tokens:
+        compact = re.sub(r"[^a-z0-9]", "", token)
+        if compact.isdigit() and 1 <= len(compact) <= 4:
+            continue
+        if compact.isalpha() and len(compact) <= 3:
+            continue
+        return False
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7425]: Fixed in 386ac6bb. Split premise components now accept bounded digit-plus-letter forms such as 221B and 12A while retaining the prior rejection rules for unrelated tokens.

File: scripts/build_data.py [hunk 32]
```diff
@@ -6732,175 +6867,1533 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        part = re.sub(r"#\s*[\w-]+", "", part)
+        address_key = raw_place_coordinate_address_key(part)
+        if address_key is None:
+            continue
+        part_tokens = raw_place_coordinate_address_token_list(address_key)
+        if raw_place_coordinate_part_has_street_intersection(
+            part_tokens,
+            part=raw_part,
+        ):
+            part_tokens = [token for token in part_tokens if token != "and"]
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 386ac6bb. Intersection comparison now removes both `and` and `at` connector tokens after intersection evidence is established; the Main St at/ampersand regression preserves the prior pin.

File: scripts/build_data.py [hunk 33]
```diff
@@ -6732,175 +6867,1533 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        len(left_google_places_ids) <= 1 and len(right_google_places_ids) <= 1
+    )
+
+    left_maps_place_token = raw_place_maps_place_token_identity(left)
+    right_maps_place_token = raw_place_maps_place_token_identity(right)
+    if (
+        left_maps_place_token is not None
+        and left_maps_place_token == right_maps_place_token
+    ):
         return True
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:6982]: Fixed in 386ac6bb. A shared maps token is accepted only when Places-ID sets are internally consistent and neither Places IDs nor legacy Google IDs conflict across rows. Regressions cover stored IDs, query_place_id values, and legacy IDs.

File: scripts/build_data.py [hunk 34]
```diff
@@ -6732,280 +6880,1772 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    if re.match(
+        r"\s*(?:c(?:\.|/)|p(?:\.\s*)?za\.?|pl\.)\s*(?=[^\W\d_])",
+        normalized_part,
+    ):
+        return True
+    street_suffix_indexes = {
+        index
+        for index, token in enumerate(tokens)
+        if index > 0 and token in RAW_PLACE_STREET_SUFFIX_TOKENS
+    }
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7396]: Fixed in 9d3b3fa. Prefix-style street markers now recognize the full `avenue` and `boulevard` forms, with regressions covering Avenue Louise and Boulevard Saint-Germain.

File: scripts/build_data.py [hunk 35]
```diff
@@ -6732,280 +6880,1772 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+    if not street_suffix_indexes:
+        return bool(numeric_indexes)
+    first_suffix_index = min(street_suffix_indexes)
+    last_suffix_index = max(street_suffix_indexes)
+    return any(
+        index < first_suffix_index or index == last_suffix_index + 1
+        for index in numeric_indexes
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7471]: Fixed in 9d3b3fa. Hash-prefixed unit tokens are now excluded from premise-number evidence using occurrence-aware indexes. The regression covers `Main St #200` as negative evidence and `200 Main St #200` as the control.

File: scripts/build_data.py [hunk 36]
```diff
@@ -6732,280 +6880,1772 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
+        left_google_places_ids & right_google_places_ids
+    ):
+        return True
 
-def raw_place_cid_identity(place: RawPlace) -> str | None:
-    return as_string(place.cid) or extract_maps_cid(place.maps_url)
+    left_primary_cid = raw_place_cid_identity(left)
+    right_primary_cid = raw_place_cid_identity(right)
+    if left_primary_cid is not None and left_primary_cid == right_primary_cid:
+        return True
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 9d3b3fa. Primary and alias CID matches now reject cross-row Google Places ID or legacy Google ID conflicts, while preserving the intended behavior when secondary IDs are missing or merely internally inconsistent.

**Comment by michaelmwu** [comment]
@codex review

File: scripts/build_data.py [hunk 37]
```diff
@@ -13336,6 +15163,7 @@ def place_selector_matches(
         ("cid", place.cid),
         ("cid", extract_maps_cid(place.maps_url)),
         *((("cid", cid_alias) for cid_alias in cid_aliases)),
+        ("gpid", extract_maps_query_place_id(place.maps_url)),
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py]: Fixed in 143f1280 (and downstream c83f04a9). `gpid:` selector values now come from the complete `raw_place_google_places_identities()` set, covering both URL query IDs and stored `ChIJ...` / `places/ChIJ...` IDs. The regression covers all bare, prefixed, and slug-qualified forms.

File: scripts/build_data.py [hunk 38]
```diff
@@ -6732,291 +6880,1850 @@ def raw_place_strong_google_identity_matches(left: RawPlace, right: RawPlace) ->
...
-        if prior_address and place_address:
-            if prior_address == place_address:
-                prior_score += 50
-            else:
-                prior_score += token_overlap_score(prior_address, place_address)
+    return bool(
+        names_are_compatible
+        and raw_place_has_durable_google_identity(left)
+        and not raw_place_has_durable_google_identity(right)
+    )
```
[the remark this answers is not in the document]
Comment (michaelmwu) [reply scripts/build_data.py:7038]: Fixed in 143f1280 (and downstream c83f04a9). The saved-list merge now has a narrowly gated name-only fallback when the refreshed row loses both durable identity and address. It requires a unique compatible name on both sides, a durable identity on the existing row, and a precise/stable prior address. A second regression proves ambiguous duplicate names remain unmatched.

**Comment by michaelmwu** [comment]
@codex review
